lib/JSONAPISerializer.js

```diff
@@ -332,6 +380,46 @@ module.exports = class JSONAPISerializer {
   }
 
   /**
+   * Serialize mixed resource object with a dynamic type resolved from data
+   *
+   * @see {@link http://jsonapi.org/format/#document-resource-objects}
+   * @method JSONAPISerializer#serializeMixedData
+   * @private
+   * @param {Object} typeOption a dynamic type options.
+   * @param {Object|Object[]} data input data.
+   * @param {Object[]} included.
+   * @return {Object|Object[]} serialized data.
+   */
+  serializeMixedData(typeOption, data, included) {
+    // Empty data
+    if (_.isEmpty(data)) {
+      // Return [] or null
+      return Array.isArray(data) ? data : null;
+    }
+
+    // Array data
+    if (Array.isArray(data)) {
+      return data.map(d => this.serializeMixedData(typeOption, d, included));
+    }
+
+    // Single data
+    // Resolve type from data (can be a string or a function deriving a type-string from each data-item)
+    const type = (typeof typeOption.type === 'function') ?
+      typeOption.type(data) :
+      _.get(data, typeOption.type);
```

Comment:
That's really only a matter of style, but I find this formatting of ternary expressions more readable:
```javascript
const type = (typeof typeOption.type === 'function')
  ? typeOption.type(data)
  : _.get(data, typeOption.type);
```
But it's just a proposal. It really depends on the way you're used to read it ;)